### PR TITLE
NMAP test upgrade - service detection

### PR DESCRIPTION
**Pull request by noursaidi**
This PR is still WIP but close to completion - submitting PR for any comments. Outstanding is just implementing CI tests and
and documentation. I'm also invesitgating if the HTTP test can be merged into this the services detection to save running two 65k port NMAP scans.

This PR maintains the current NMAP functionality and extends it by also check for disallowed services on all ports ([test document here](https://docs.google.com/document/d/1wLDV5uvJrYtWem0xeu22uLnS4Dh-1TZggvFBDscBla4/edit)). in summary:
- runs a nmap scan against all ports with service detection option (-sV) for TCP ports
- runs an nmap scan against ports 69,161,162 for UDP and any other UDP port in module_config.json
- For each TCP services(pop, telnet, ftp, imap, smtp), check if the service has been found in the NMAP scan or it's default port was open 
- For UDP services (tftp, SNMP), check if the default port was open
- Output results for each test item (security.service.X)

The existing nmap scan has been retained for now (expected to be used within ATA) and is selected with a `"services_scan": false`

### CODE REVIEW

**Comment by noursaidi** [comment]
@grafnu @pisuke do you have any comments or questions  on this PR? Except for any on small CI fixes which I'm working my way though (test_mud may cause me some issues but I'm diagnosing this one)

File: docker/include/bin/start_faux [hunk 1]
```diff
@@ -134,7 +134,44 @@ echo Faux device at $local_ip bcast $broadcast_ip
 # Setup fake telnet listener to trigger port-scan failure
 if [ -n "${options[telnet]}" ]; then
     echo Enabling mock telnet server...
-    (while true; do echo Telnet `hostname`; nc -nvlt -p 23 -e `which hostname`; done) &
+    while true; do echo -e "Telnet $(hostname)" | nc -l -w 1 23; done)&
```
Comment (grafnu) [inline docker/include/bin/start_faux:137]: seems like this would be a syntax error?  ... stray )?

File: docker/include/bin/start_faux [hunk 2]
```diff
@@ -134,7 +134,44 @@ echo Faux device at $local_ip bcast $broadcast_ip
...
+fi
+
+if [ -n "${options[oddservices]}" ]; then
+    echo Running services on non standard ports and open default ports 
+
+    echo Starting FTP 21514 and open default 20,21
+    nc -nvlt -p 20&
```
Comment (grafnu) [inline docker/include/bin/start_faux:144]: space before & (everywhere)

File: docker/include/bin/start_faux [hunk 3]
```diff
@@ -134,7 +134,44 @@ echo Faux device at $local_ip bcast $broadcast_ip
...
+    nc -nvlt -p 587&
+    (while true; do echo -e "220 $(hostname) ESMTP Postfix (Ubuntu)" | nc -l -w 1 1256; done)&
+
+    echo Starting IMAP 5361 and open default ports 143, 993
+    nc -nvlt -p 143&
+    nc -nvlt -p 993&
+    (while true; do echo -e "* OK [CAPABILITY IMAP4rev1 LITERAL+ SASL-IR LOGIN-REFERRALS ID ENABLE IDLE STARTTLS AUTH=PLAIN] Dovecot (Ubuntu) ready.\r\n" \
+        | nc -l -w 1 5361; done)&
+
+    #echo Starting TELNET and default port 23
```
Comment (grafnu) [inline docker/include/bin/start_faux:160]: restore or remove

File: bin/build_daq_container [hunk 4]
```diff
@@ -15,3 +15,5 @@ echo Starting $image_name build...
 docker build . -f docker/modules/Dockerfile.daq --build-arg GIT_URL -t daqf/$image_name
 
 echo Done with $image_name build.
+
+docker build . -f subset/pentests/nmap/Dockerfile.test_nmap -t daqf/nmap
```
Comment (grafnu) [inline bin/build_daq_container:19]: missing newline

File: subset/pentests/nmap/nmap_results_reader.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        Returns:
+            Dictionary of ports which are considered open and are found to
+            have a product name matching the supplied regex
+
+        """
+        return {k: v for k, v in self.ports.items() if
+                v.protocol == protocol and re.search(product, v.product)}
+
+
+if __name__ == "__main__":
```
Comment (grafnu) [inline subset/pentests/nmap/nmap_results_reader.py:114]: what's this for?

**Review by grafnu — COMMENTED** [review]
We shouldn't worry about mud at all -- so if there's an easy way to remove/deprecate then that would be fine. The current mud stuff is all very very very very provisional and doesn't really do anything, I think.

File: subset/pentests/nmap/nmap_results_reader.py (hunk 5 above)
Comment (noursaidi) [reply subset/pentests/nmap/nmap_results_reader.py:114]: redundant in this file - removed

File: bin/build_daq_container (hunk 4 above)
Comment (noursaidi) [reply bin/build_daq_container:19]: This was an accidental file change - modification removed

File: docker/include/bin/start_faux (hunk 2 above)
Comment (noursaidi) [reply docker/include/bin/start_faux:144]: Added spaces

File: docker/include/bin/start_faux (hunk 3 above)
Comment (noursaidi) [reply docker/include/bin/start_faux]: Removed and existing telnet faux option reinstated

**Comment by noursaidi** [comment]
Comments addressed - CI tests should be functional

File: testing/test_mud.sh [hunk 6]
```diff
@@ -7,6 +7,12 @@ echo MUD Tests >> $TEST_RESULTS
 rm -rf inst/tmp_site && mkdir -p inst/tmp_site
 cp resources/setups/baseline/report_template.md inst/tmp_site/
 
+mkdir -p local/site
+# set nmap module to legacy nmap port based scan
+cat resources/setups/common/base_config.json \
+| jq '.modules .nmap .services_scan = false' \
```
Comment (grafnu) [inline testing/test_mud.sh:13]: nit -- indent hanging lines 4 spaces

File: testing/test_mud.sh [hunk 7]
```diff
@@ -17,6 +23,8 @@ echo %%%%%%%%%%%%%%%%%%%%%% Default MUD | tee -a $TEST_RESULTS
 cmd/run -s interfaces.faux.opts=telnet device_specs=resources/device_specs/simple.json
 echo DAQ result code $? | tee -a $TEST_RESULTS
 cat inst/result.log | tee -a $TEST_RESULTS
+head -20 inst/run-*/nodes/*/activate.log
```
Comment (grafnu) [inline testing/test_mud.sh:26]: potentially extra debugging statements

**Review by grafnu — APPROVED** [review]

**Pull request merged**